### ISSUE TITLE
HW Accel issue on rk3588

### ISSUE THREAD

**Issue by System64fumo** [FEX-Emu/FEX#4236]
Hello, I've recently switched from box64 to fex since i can't get steam/games working there.
When running steam i get the following harmless errors: (Steam runs fine but software rendered)
```
MESA: error: failed to mmap the LATEST_FLUSH_ID register (err=22)
glx: failed to create dri3 screen
failed to load driver: rockchip
MESA: error: failed to mmap the LATEST_FLUSH_ID register (err=22)
glx: failed to create dri3 screen
failed to load driver: rockchip
ATTENTION: default value of option mesa_glthread overridden by environment.
MESA: error: failed to mmap the LATEST_FLUSH_ID register (err=22)
glx: failed to create dri3 screen
failed to load driver: rockchip
MESA: error: failed to mmap the LATEST_FLUSH_ID register (err=22)
glx: failed to create dri3 screen
failed to load driver: rockchip
ATTENTION: default value of option mesa_glthread overridden by environment.
MESA: error: failed to mmap the LATEST_FLUSH_ID register (err=22)
glx: failed to create dri3 screen
failed to load driver: rockchip
MESA: error: failed to mmap the LATEST_FLUSH_ID register (err=22)
glx: failed to create dri3 screen
failed to load driver: rockchip
```
I've confirmed that /lib/dri/rockchip_dir.so exists in both the aarch64 root and x86_64 root
I've also checked using `glxinfo -B` and confirmed that hw accel is working
Even tried running a game, All works fine.

Any ideas?

Edit: I'm running linux 6.13.0-rc3 (Panthor GPU driver) with fex 2410-1 on armtix linux on an Orange Pi 5 Plus

**Comment by System64fumo** [comment]
Also on topic, Not sure if it's even possible at such an early stage but is it possible to use vulkan? (With panthor)

On the aarch64 root i can run `PAN_I_WANT_A_BROKEN_VULKAN_DRIVER=1 vkcube` and get:
```
Selected WSI platform: xcb
WARNING: panvk is not a conformant Vulkan implementation, testing use only.
Selected GPU 0: Mali-G610 (Panfrost), type: IntegratedGpu
```

But on the x86_64 one it's llvmpipe.

Again, Totally unsure if it's possible at such an early stage, Just curios

**Comment by Sonicadvance1** [comment]
```
#define DRM_PANTHOR_USER_MMIO_OFFSET_32BIT	(1ull << 43)
#define DRM_PANTHOR_USER_MMIO_OFFSET_64BIT	(1ull << 56)
#define DRM_PANTHOR_USER_MMIO_OFFSET		(sizeof(unsigned long) < 8 ? \
						 DRM_PANTHOR_USER_MMIO_OFFSET_32BIT : \
						 DRM_PANTHOR_USER_MMIO_OFFSET_64BIT)
#define DRM_PANTHOR_USER_FLUSH_ID_MMIO_OFFSET	(DRM_PANTHOR_USER_MMIO_OFFSET | 0)

```

Looks like the Panthor devs messed up and bound their FLUSH_ID MMIO offset to something that changes between 32-bit and 64-bit processes. Because FEX always presents as a 64-bit process, this `DRM_PANTHOR_USER_FLUSH_ID_MMIO_OFFSET` should always be the 64-bit one, but because Steam is 32-bit it will end up using `DRM_PANTHOR_USER_MMIO_OFFSET_32BIT`.
This gets passed directly to mmap (or mmap2 in the case of 32-bit) for the file offset and breaks things.

They should have bound this to an ioctl, but since they decided to take this roundabout approach, it looks like Panthor will just be broken when running 32-bit processes. We don't have any way to track DRM FDs through to mmap (and we probably don't want to eat the overhead that incurs?)

Not sure why x86-64 vkcube isn't working for you though, maybe they made other mistakes elsewhere with their cross-architecture cleanliness.

**Comment by System64fumo** [comment]
Ah, So it's a panthor issue, Got it :+1:
Doubt this would get fixed upstream, Should i bring it up though?

And as for vkcube
It's not just vkcube, Any vulkan program fails and uses the sw renderer instead.

Perhaps it's just because the vulkan driver is still broken and not finished, Thanks though.

**Comment by Sonicadvance1** [comment]
> Ah, So it's a panthor issue, Got it 👍 Doubt this would get fixed upstream, Should i bring it up though?

Maybe? I haven't talked with their developers in a couple years. Not sure how much they care about changing their UAPI for x86 compatibility.

> 
> And as for vkcube It's not just vkcube, Any vulkan program fails and uses the sw renderer instead.
> 
> Perhaps it's just because the vulkan driver is still broken and not finished, Thanks though.

**Comment by System64fumo** [comment]
Very well, I'll see if i can get in touch with any of the panthor devs to help resolve this :+1:

Thank you for your time!

**Comment by amazingfate** [comment]
After updating to mesa 25.2.3 with panfrost driver inside rootfs and fix of issue https://gitlab.freedesktop.org/mesa/mesa/-/issues/12364, I still can't get steam gpu accelerated:
```
09/30 20:57:57 Init: Installing breakpad exception handler for appid(steamsysinfo)/version(1757650573)/tid(8825)
Running query: 1 - GpuTopology
MESA: error: failed to mmap the LATEST_FLUSH_ID register (err=22)
CVulkanTopology: failed to get physical device count
Failed to query vulkan gpu topology

Failed to query vulkan gpu topology
Response: 
Exit code: -2
/bin/bash: warning: setlocale: LC_ALL: cannot change locale (zh_CN.UTF-8)
steamwebhelper.sh[8846]: Starting steamwebhelper under bootstrap steamrt steam runtime via: /home/jfliu/.local/share/Steam/steamrt64/steam-runtime-steamrt/_v2-entry-point
steamwebhelper.sh[8846]: Starting steamwebhelper with steamrt steam runtime at /home/jfliu/.local/share/Steam/steamrt64/steam-runtime-steamrt/_v2-entry-point
Steam Runtime Launch Service: starting steam-runtime-launcher-service
Steam Runtime Launch Service: steam-runtime-launcher-service is running pid 8904
bus_name=com.steampowered.PressureVessel.LaunchAlongsideSteam
MESA: error: failed to mmap the LATEST_FLUSH_ID register (err=22)
vkEnumeratePhysicalDevices failed, unable to init and enumerate GPUs with Vulkan.
BInit - Unable to initialize Vulkan!
```
There is no glx dri error now. Is vulkan necessary for steam?

**Comment by Sonicadvance1** [comment]
Maybe? I haven't tried on a non-vulkan system in a long time.
Looks like either your version of mesa is too old to include https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/34573 this PR.
Or your kernel is too old to support the new ioctl that is needed for that.

**Comment by amazingfate** [comment]
> Maybe? I haven't tried on a non-vulkan system in a long time.
> Looks like either your version of mesa is too old to include https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/34573 this PR.
> Or your kernel is too old to support the new ioctl that is needed for that.

I'm using kernel 6.16.4 and mesa 25.2.3, both new enough, and mesa does have included that PR.

**Comment by macromorgan** [comment]
I found a fix for this, but it's a bit of a pain in the butt... You'll need to compile your own Mesa for use in the chroot and that Mesa will need Panfrost enabled for amd64 builds.

**Comment by System64fumo** [comment]
I don't know why i closed this a while back, Whoops!
No this isn't resolved yet i recently tried FEX again on my rk3588 board, Still same issue, iirc i did try using a custom build of mesa with panfrost enabled and that didn't do it?
Could have changed since then though.

**Comment by macromorgan** [comment]
I'm using Kisak's branch which is currently something like 26.0.6 on Ubuntu 26.04 on my rk3588 device (an Indiedroid Nova). Prior to using Kisak's branch on the host I was getting crashes. I also upgraded the rootfs to the same version (26.0.6) but that didn't resolve my issue. I then recompiled Kisak's version for amd64 adding panfrost as amd64 drivers, entered the Fex chroot with `chroot.py chroot`, used dpkg -i *deb to install my Kisak drivers, and it started to work.

I've asked Kisak if he can start including Panfrost in his amd64 and i386 builds as a result.
